cloud/scope/control_plane.go

```diff
@@ -0,0 +1,594 @@
+			return errors.Wrap(err, "failed to get kubeconfig secret")
+		}
+
+		if createErr := s.createCAPIKubeconfigSecret(
+			ctx,
+			okeCluster,
+			clusterRef,
+		); createErr != nil {
+			return fmt.Errorf("creating kubeconfig secret: %w", createErr)
+		}
+	} else if updateErr := s.updateCAPIKubeconfigSecret(ctx, configSecret, okeCluster); updateErr != nil {
+		return fmt.Errorf("updating kubeconfig secret: %w", updateErr)
+	}
+
+	// Set initialized to true to indicate the kubconfig has been created
+	s.OCIManagedControlPlane.Status.Initialized = true
+
+	return nil
+}
+
+func (s *ControlPlaneScope) updateCAPIKubeconfigSecret(ctx context.Context, configSecret *corev1.Secret, okeCluster *oke.Cluster) error {
+	data, ok := configSecret.Data[secret.KubeconfigDataName]
+	if !ok {
+		return errors.Errorf("missing key %q in secret data", secret.KubeconfigDataName)
+	}
+
+	config, err := clientcmd.Load(data)
+	if err != nil {
+		return errors.Wrap(err, "failed to convert kubeconfig Secret into a clientcmdapi.Config")
+	}
+
+	token, err := s.BaseClient.GenerateToken(ctx, *okeCluster.Id)
+	if err != nil {
+		return fmt.Errorf("generating presigned token: %w", err)
+	}
+
+	if err != nil {
+		return fmt.Errorf("generating presigned token: %w", err)
+	}
+
+	userName := getKubeConfigUserName(*okeCluster.Name, false)
+	config.AuthInfos[userName].Token = token
+
+	out, err := clientcmd.Write(*config)
+	if err != nil {
+		return errors.Wrap(err, "failed to serialize config to yaml")
+	}
+
+	configSecret.Data[secret.KubeconfigDataName] = out
+
+	err = s.client.Update(ctx, configSecret)
+	if err != nil {
+		return fmt.Errorf("updating kubeconfig secret: %w", err)
+	}
+
+	return nil
+}
+
+func (s *ControlPlaneScope) buildImagePolicyConfig() *oke.CreateImagePolicyConfigDetails {
+	if s.OCIManagedControlPlane.Spec.ImagePolicyConfig != nil {
+		policyDetails := oke.CreateImagePolicyConfigDetails{
+			IsPolicyEnabled: s.OCIManagedControlPlane.Spec.ImagePolicyConfig.IsPolicyEnabled,
+		}
+		if len(s.OCIManagedControlPlane.Spec.ImagePolicyConfig.KeyDetails) > 0 {
+			keys := make([]oke.KeyDetails, 0)
+			for _, key := range s.OCIManagedControlPlane.Spec.ImagePolicyConfig.KeyDetails {
+				keys = append(keys, oke.KeyDetails{
+					KmsKeyId: key.KmsKeyId,
+				})
+			}
+		}
+		return &policyDetails
+	}
+	return nil
+}
+
+func (s *ControlPlaneScope) buildUpdateImagePolicyConfig() *oke.UpdateImagePolicyConfigDetails {
+	if s.OCIManagedControlPlane.Spec.ImagePolicyConfig != nil {
+		policyDetails := oke.UpdateImagePolicyConfigDetails{
+			IsPolicyEnabled: s.OCIManagedControlPlane.Spec.ImagePolicyConfig.IsPolicyEnabled,
+		}
+		if len(s.OCIManagedControlPlane.Spec.ImagePolicyConfig.KeyDetails) > 0 {
+			keys := make([]oke.KeyDetails, 0)
+			for _, key := range s.OCIManagedControlPlane.Spec.ImagePolicyConfig.KeyDetails {
+				keys = append(keys, oke.KeyDetails{
+					KmsKeyId: key.KmsKeyId,
+				})
+			}
+		}
+		return &policyDetails
+	}
+	return nil
+}
```

Comment:
Since `buildUpdateImagePolicyConfig` and `buildImagePolicyConfig` are almost identical functionally is it possible to create a method that passes in the policy details so we don't have to maintain two sets of logic here?

Reply:
fixed